graylog2-server/src/main/java/org/graylog2/migrations/V20220623125450_AddJobTypeToJobTrigger.java

```diff
@@ -0,0 +1,121 @@
+/*
+ * Copyright (C) 2020 Graylog, Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the Server Side Public License, version 1,
+ * as published by MongoDB, Inc.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * Server Side Public License for more details.
+ *
+ * You should have received a copy of the Server Side Public License
+ * along with this program. If not, see
+ * <http://www.mongodb.com/licensing/server-side-public-license>.
+ */
+package org.graylog2.migrations;
+
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.google.auto.value.AutoValue;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.Streams;
+import com.mongodb.client.AggregateIterable;
+import com.mongodb.client.MongoCollection;
+import com.mongodb.client.model.Aggregates;
+import com.mongodb.client.model.Filters;
+import com.mongodb.client.model.Projections;
+import com.mongodb.client.model.UpdateOneModel;
+import com.mongodb.client.model.Updates;
+import org.bson.Document;
+import org.graylog.scheduler.DBJobDefinitionService;
+import org.graylog.scheduler.DBJobTriggerService;
+import org.graylog.scheduler.JobTriggerDto;
+import org.graylog2.database.MongoConnection;
+import org.graylog2.plugin.cluster.ClusterConfigService;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.inject.Inject;
+import java.time.ZonedDateTime;
+import java.util.Arrays;
+import java.util.List;
+import java.util.stream.Collectors;
+
+public class V20220623125450_AddJobTypeToJobTrigger extends Migration {
+    private static final Logger LOG = LoggerFactory.getLogger(V20220623125450_AddJobTypeToJobTrigger.class);
+
+
+    private final MongoConnection mongoConnection;
+
+    private final ClusterConfigService clusterConfigService;
+
+    @Inject
+    public V20220623125450_AddJobTypeToJobTrigger(MongoConnection mongoConnection, ClusterConfigService clusterConfigService) {
+        this.mongoConnection = mongoConnection;
+        this.clusterConfigService = clusterConfigService;
+    }
+
+    @Override
+    public ZonedDateTime createdAt() {
+        return ZonedDateTime.parse("2022-06-23T12:54:50Z");
+    }
+
+    @Override
+    public void upgrade() {
+        final MigrationCompleted migrationCompleted = clusterConfigService.getOrDefault(MigrationCompleted.class, MigrationCompleted.createEmpty());
+
+        // Run this more than once in case old cluster nodes are still writing triggerDTOs with empty types
+        if (migrationCompleted.executionCount() >= 3) {
+            return;
+        }
```

Comment:
I think this running this three times is quite arbitrary. :smile: How about removing the `MigrationCompleted` value and always adding a quick guard query, and only running the migration if there are any triggers that don't have the `job_definition_type` field?

Reply:
I agree that it is, but I was afraid that this might be expensive. After all, I've witnessed setups with millions of notification triggers already

Reply:
I'm not a mongodb expert, but I think I've read somewhere that count might be expensive, even though the field has an index.
e.g.: https://www.mongodb.com/community/forums/t/why-is-count-so-slow/110389

However, this reminded me to add an index to `FIELD_JOB_DEFINITION_TYPE` as well now :+1: